### PR TITLE
Webhook framework for cloud controller manager

### PR DESCRIPTION
**Pull request by nckturner**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
We need to allow cloud providers to host webhooks in their cloud controller managers in order to replace the persistent volume labeller admission controller that is currently present in the API server.  

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

See example usage here: https://github.com/kubernetes/cloud-provider-aws/pull/321.  I'll add a similar example to the example in cmd/cloud-controller-manager.go as well.

#### Does this PR introduce a user-facing change?

```release-note
Added the ability to host webhooks in the cloud controller manager.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: [KEP-2699](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2699-add-webhook-hosting-to-ccm)
```

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/app/webhooks.go:23]: Interesting thought, do you mean that the framework with a new interface when a new use case would be added?

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/app/webhooks.go]: Yeah and start with two known use-cases, PersistentVolumes and Services.The enablement of the webhook could also be done by the provider implementation, similar to what we already do for enabling "Instances", "LoadBalancers", etc

File: staging/src/k8s.io/cloud-provider/options/options.go
Comment (cheftako) [reply staging/src/k8s.io/cloud-provider/options/options.go:79]: My general feeling is that is will be similar to the CCM's controller list. The general webhook list will probably (at least initially) just have the PV webhook in it. However it should be possible for cloud providers to add there own ala our example with controllers of adding the NodeIPAMController. I would hope we could use the same pattern to keep things consistent for people needing to do this.

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/app/webhooks.go:60]: The whole thing seems like the webhook. Would it be better to call this something like Admit or AdmissionHandler?

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/app/webhooks.go:68]: Should we have logic here to "skip" adding any webhooks which have been disabled through the CLI?

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/app/controllermanager.go:169]: Should we have a feature gate blocking the enablement of the webhook handlers?

File: staging/src/k8s.io/cloud-provider/app/testing/testserver.go
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/app/testing/testserver.go:136]: skip it?

**Comment by aojea** [comment]
can you add the reference to the KEP on the description section?
 https://github.com/kubernetes/enhancements/issues/2699

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/app/webhooks.go]: Walter and I discussed this and decided that the AdmissionRequest and AdmissionResponse model is more useful and extensible, but let's continue discussion in the sig if necessary.

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/app/webhooks.go:118]: We should probably wrap these requests in metrics. I would think we would want latency and response code metrics by canoncial webhook name.

**Comment by leonardpahlke** [comment]
Hi, Thanks for raising the [exception request](https://groups.google.com/a/kubernetes.io/g/release-team/c/ZODxaU5lKeA). The exception request has been [approved](https://groups.google.com/a/kubernetes.io/g/release-team/c/ZODxaU5lKeA/m/DfhT1E6gAAAJ) by the 1.26 release team and your updated deadline to make any changes to your KEP is 18:00 PST Thursday 13 October 2022.
/milestone v1.26

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/app/controllermanager.go:99]: For readability it would be nice if constructed and passed in allWebhooks and disabledByDefaultWebhooks variables for the new parameters.

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/app/controllermanager.go:126]: Again for readability.

File: staging/src/k8s.io/cloud-provider/config/types.go
Comment (roycaihw) [inline staging/src/k8s.io/cloud-provider/config/types.go:98]: nit: you documented "all webhooks" and "disabled by default webhooks" in `AddFlag`. Is "enabled by default webhooks" the diff of those two groups?

**Review by roycaihw — COMMENTED** [review]
/approve
for api-rules exception

**Review by dims — CHANGES_REQUESTED** [review]
https://github.com/kubernetes/kubernetes/pull/108838/files#diff-71e3b98f9a6bbf5b8421e26a7ba0c079f397cd8d49abacdad943c66a4f44f03dR176

this should be v1.27 :)

File: staging/src/k8s.io/cloud-provider/app/controllermanager.go
Comment (dims) [inline staging/src/k8s.io/cloud-provider/app/controllermanager.go:190]: this line can go into the `if` condition?

**Comment by dims** [comment]
hack/, pkg/features, test/ changes look good. a couple of nits inline

/approve

File: staging/src/k8s.io/cloud-provider/config/types.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/config/types.go:98]: Yes, that is correct

**Comment by nckturner** [comment]
@dims done

**Comment by dims** [comment]
/assign @liggitt 

Jordan, a few lines in `api/api-rules/violation_exceptions.list` need your approval. thanks!

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/app/webhooks.go:198]: does this ensure the version sent to the webhook was actually a v1 AdmissionReview?

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/app/webhooks.go:53]: we shouldn't need to add admissionregistration, this is only dealing with AdmissionReview objects, right?

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/app/webhooks.go:55]: can we limit this to supporting v1 AdmissionReview objects?

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/app/webhooks.go:206]: I'm not sure what this is doing... this sets `review` to an object with a nil Request, which means we hit line 210 below and just return an error

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (liggitt) [reply staging/src/k8s.io/cloud-provider/app/webhooks.go:198]: staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go might be a good pattern to copy with some defensive limiting of the body size we read in, and more efficient writing of the response

File: staging/src/k8s.io/cloud-provider/options/options.go
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/options/options.go:178]: can WebhookServing be nil? will this panic?

File: staging/src/k8s.io/cloud-provider/options/options.go
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/options/options.go:261]: can WebhookServing be nil?

File: staging/src/k8s.io/cloud-provider/options/webhook.go
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/options/webhook.go:93]: this defaults to publicly serving on all interfaces? should the consuming CCM pass in the port they want and the default addresses, rather than assuming all consumers that have webhooks want them served publicly?

File: staging/src/k8s.io/cloud-provider/options/webhook.go
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/options/webhook.go:36]: should this also be referenced from pkg/cluster/ports/ports.go where we already declare CloudControllerManagerPort?

File: staging/src/k8s.io/cloud-provider/options/options.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/options/options.go:178]: Not if the consuming cloud-controller-manager is using the provided functions to initialize their options, like `NewCloudControllerManagerOptions`, but since they could be overriding this I'll add a nil check here.

File: staging/src/k8s.io/cloud-provider/options/options.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/options/options.go:261]: Yeah in general the ones I've seen are, but obviously I can't know about all of them so I'll add a nil check.

File: staging/src/k8s.io/cloud-provider/options/webhook.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/options/webhook.go:36]: Yes, I'll add a reference from there to here in addition to the other CCM port, which is for the healthz and metrics server.

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/app/webhooks.go:200]: Fixed to return error explicitly.

File: staging/src/k8s.io/cloud-provider/options/webhook.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/options/webhook.go:93]: That's a good idea :+1:

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/app/webhooks.go:191]: > does this ensure the version sent to the webhook was actually a v1 AdmissionReview?

Addressed this comment

> staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go might be a good pattern to copy with some defensive limiting of the body size we read in, and more efficient writing of the response

Can this be done in a follow up, in order to be sure we hit code freeze?

File: staging/src/k8s.io/cloud-provider/options/webhook.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/options/webhook.go:93]: I've added a ProviderDefaults struct which can be passed in by the CCM consumer to set the override the default flag values (still can be overridden by a set flag).

File: staging/src/k8s.io/cloud-provider/options/options.go
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/options/options.go:175]: I don't have a good sense for how this library is used, but if webhook / webhookserving could be nil, need to guard the AddFlags and Validate calls as well

**Comment by liggitt** [comment]
/approve
for the API lint file

/hold for nil checks on validate/addflags (https://github.com/kubernetes/kubernetes/pull/108838#discussion_r1136153686) and review of updated impl from cloud-provider reviewer

File: staging/src/k8s.io/cloud-provider/options/options.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/options/options.go:196]: done.

**Comment by liggitt** [comment]
/hold cancel

impl needs review/lgtm by cloud-provider approve

File: staging/src/k8s.io/cloud-provider/options/webhook.go
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/options/webhook.go:160]: Can we make sure we don't reuse one of the other CCM ports? Concerned the two SecureServingInfo will conflict.

Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/options/webhook.go:159]: Added a validation check to `func (o *CloudControllerManagerOptions) Validate(...) ...`.

**Comment by nckturner** [comment]
@cheftako for final LGTM

File: staging/src/k8s.io/cloud-provider/app/webhooks.go
Comment (nckturner) [reply staging/src/k8s.io/cloud-provider/app/webhooks.go:191]: Spoke offline about this, and we're going to do this in a followup

**Comment by marosset** [comment]
@dims can you please update your review state to approved so this PR can merge?

**Pull request merged**
